src/Workspaces/Core/Portable/FindSymbols/FindReferences/DependentTypeFinder.cs

```diff
@@ -128,88 +133,212 @@ internal static partial class DependentTypeFinder
+                // started with a metadata type (i.e. a source type could not have a descendant type found in metadata,
+                // but a metadata type could have descendant types in source and metadata).
+                if (searchInMetadata)
+                {
+                    using var _ = GetSymbolSet(out var tempBuffer);
+
+                    await AddDescendantMetadataTypesInProjectAsync(tempBuffer, project).ConfigureAwait(false);
+
+                    // Add all the matches we found to the result set.
+                    AssertContents(tempBuffer, assert: s_isInMetadata, "Found type was not from metadata");
+                    AddRange(tempBuffer, result);
+
+                    // Now, if we're doing a transitive search, add these found types to the 'current' sets we're
+                    // searching for more results for. These will then be used when searching for more types in the next
+                    // project (which our caller controls).
+                    if (transitive)
+                    {
+                        AddRange(tempBuffer, currentMetadataTypes, shouldContinueSearching);
+                        AddRange(tempBuffer, currentSourceAndMetadataTypes, shouldContinueSearching);
+                    }
+                }
+
                 {
-                    await DescendInheritanceTreeInProjectAsync(
-                        searchInMetadata, result,
-                        currentMetadataTypes, currentSourceAndMetadataTypes,
-                        project,
-                        typeMatches,
-                        shouldContinueSearching,
-                        transitive, cancellationToken).ConfigureAwait(false);
+                    using var _ = GetSymbolSet(out var tempBuffer);
+
+                    // Now search the project and see what source types we can find.
+                    await AddDescendantSourceTypesInProjectAsync(tempBuffer, project).ConfigureAwait(false);
+
+                    // Add all the matches we found to the result set.
+                    AssertContents(tempBuffer, assert: s_isInSource, "Found type was not from source");
+                    AddRange(tempBuffer, result);
+
+                    // Now, if we're doing a transitive search, add these types to the currentSourceAndMetadataTypes
+                    // set. These will then be used when searching for more types in the next project (which our caller
+                    // controls).  We don't have to add this to currentMetadataTypes since these will all be
+                    // source types.
+                    if (transitive)
+                        AddRange(tempBuffer, currentSourceAndMetadataTypes, shouldContinueSearching);
                 }
             }
 
-            return result.ToImmutableArray();
-        }
+            async Task AddDescendantSourceTypesInProjectAsync(SymbolSet result, Project project)
```

Comment:
just a move.

Reply:
unchanged, except that it has a reduced number of arguments due to being a nested function now.